mmv1/third_party/terraform/services/artifactregistry/data_source_artifact_registry_docker_image.go

```diff
@@ -0,0 +1,331 @@
+				Description: `The URI to access the image.`,
+			},
+			"tags": {
+				Type:        schema.TypeList,
+				Computed:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Description: `All tags associated with the image.`,
+			},
+			"image_size_bytes": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: `Calculated size of the image in bytes.`,
+			},
+			"media_type": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: `Media type of this image.`,
+			},
+			"upload_time": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: `The time, as a RFC 3339 string, the image was uploaded.`,
+			},
+			"build_time": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: `The time, as a RFC 3339 string, this image was built.`,
+			},
+			"update_time": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: `The time, as a RFC 3339 string, this image was updated.`,
+			},
+		},
+	}
+}
+
+func DataSourceArtifactRegistryDockerImageRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*transport_tpg.Config)
+	userAgent, err := tpgresource.GenerateUserAgentString(d, config.UserAgent)
+	if err != nil {
+		return err
+	}
+
+	project, err := tpgresource.GetProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	var res DockerImage
+
+	// check that only one of digest or tag is set
+	_, hasDigest := d.GetOk("digest")
+	tag, hasTag := d.GetOk("tag")
+
+	if hasDigest && hasTag {
+		return fmt.Errorf("only one of tag or digest can be set")
+	}
+
+	if hasDigest {
+		// fetch image by digest
+		// https://cloud.google.com/artifact-registry/docs/reference/rest/v1/projects.locations.repositories.dockerImages/get
+		imageUrlSafe := url.QueryEscape(d.Get("image").(string))
+		urlRequest, err := tpgresource.ReplaceVars(d, config, fmt.Sprintf("{{ArtifactRegistryBasePath}}projects/{{project}}/locations/{{region}}/repositories/{{repository}}/dockerImages/%s@{{digest}}", imageUrlSafe))
+		if err != nil {
+			return fmt.Errorf("Error setting api endpoint")
+		}
+
+		resGet, err := transport_tpg.SendRequest(transport_tpg.SendRequestOptions{
+			Config:    config,
+			Method:    "GET",
+			Project:   project,
+			RawURL:    urlRequest,
+			UserAgent: userAgent,
+		})
+		if err != nil {
+			return err
+		}
+
+		res = convertResponseToStruct(resGet)
+	} else {
+		// fetch the list of images, ordered by update time
+		// https://cloud.google.com/artifact-registry/docs/reference/rest/v1/projects.locations.repositories.dockerImages/list
+		urlRequest, err := tpgresource.ReplaceVars(d, config, "{{ArtifactRegistryBasePath}}projects/{{project}}/locations/{{region}}/repositories/{{repository}}/dockerImages")
+		if err != nil {
+			return fmt.Errorf("Error setting api endpoint")
+		}
+
+		urlRequest, err = transport_tpg.AddQueryParams(urlRequest, map[string]string{"orderBy": "update_time desc"})
```

Comment:
Just verifying, is this sorting going to be relevant for all (or most) use cases?

Reply:
It's more of an optimization: since there can be only one image with a particular label, it's most likely going to be one that was updated recently.

Reply:
Ah, that makes sense.

I think I was also asking about this case (from your doc change): `If a digest or tag is not provided, then the last updated version of the image will be fetched.`. Is the most recently updated going to typically be what users want?

Reply:
@roaks3 Yes, I think that's a reasonable expectation that users have.  It's also the behavior of other API calls, for example when calling the [image of a family](https://cloud.google.com/compute/docs/reference/rest/v1/images#getfromfamily), the latest image is returned.